File: 2724-convert-an-array-into-a-2d-array-with-conditions/convert-an-array-into-a-2d-array-with-conditions.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


class Solution {
   
        





    public List<List<Integer>> findMatrix(int[] nums) {
        Map<Integer, Integer> mp = new HashMap<>();
        for (int num : nums) {
            mp.put(num, mp.getOrDefault(num, 0) + 1);
        }
        
        List<List<Integer>> ans = new ArrayList<>();
        while (!mp.isEmpty()) {
            List<Integer> li = new ArrayList<>();
            
            // iterating on copy to avoid getting ConcurrentModificationException
            Map<Integer, Integer> mpCopy = new HashMap<>(mp);
            for (int num : mpCopy.keySet()) {
                li.add(num);
                mp.put(num, mp.get(num) - 1);
                if (mp.get(num) == 0) {
                    mp.remove(num);
                }
            }
            
            ans.add(li);
        }
        
        return ans;
    }
}

 

File: 2724-convert-an-array-into-a-2d-array-with-conditions/convert-an-array-into-a-2d-array-with-conditions.py
class Solution:
    def findMatrix(self, nums: List[int]) -> List[List[int]]:

        mp ={}
        for num in nums:
            mp[num] = mp.get(num,0)+1
        ans= []
        while mp:
            li =[]
           
            for num in mp.copy():
                li.append(num)
                mp[num]-=1
                if mp[num] ==0:
                    mp.pop(num)

            ans.append(li)
    
        return ans 

        